src/mpi/romio/adio/ad_daos/ad_daos_io.c

```diff
@@ -0,0 +1,239 @@
+    iod->arr_nr = 1;
+    rg->rg_len = len;
+    rg->rg_idx = offset;
+    iod->arr_rgs = rg;
+
+#ifdef ADIOI_MPE_LOGGING
+    MPE_Log_event(ADIOI_MPE_write_a, 0, NULL);
+#endif
+
+#ifdef D_PRINT_IO
+    int mpi_rank;
+
+    MPI_Comm_rank(fd->comm, &mpi_rank);
+    printf("(%d) CONTIG IO OP %d, Off %llu, Len %zu\n", mpi_rank, flag, offset, len);
+#endif
+
+    if (flag == DAOS_WRITE) {
+        ret = daos_array_write(cont->oh, DAOS_TX_NONE, iod, sgl, NULL,
+                               (request ? &aio_req->daos_event : NULL));
+        if (ret != 0) {
+            PRINT_MSG(stderr, "daos_array_write() failed with %d\n", ret);
+            *error_code = ADIOI_DAOS_err(myname, cont->obj_name, __LINE__, ret);
+            return;
+        }
+    } else if (flag == DAOS_READ) {
+        ret = daos_array_read(cont->oh, DAOS_TX_NONE, iod, sgl, NULL,
+                              (request ? &aio_req->daos_event : NULL));
+        if (ret != 0) {
+            PRINT_MSG(stderr, "daos_array_read() failed with %d\n", ret);
+            *error_code = ADIOI_DAOS_err(myname, cont->obj_name, __LINE__, ret);
+            return;
+        }
+    }
+#ifdef ADIOI_MPE_LOGGING
+    MPE_Log_event(ADIOI_MPE_write_b, 0, NULL);
+#endif
+
+    if (file_ptr_type == ADIO_INDIVIDUAL) {
+        fd->fp_ind += len;
+    }
+
+    fd->fp_sys_posn = offset + len;
+
+  done:
+#ifdef HAVE_STATUS_SET_BYTES
+    if (status)
+        MPIR_Status_set_bytes(status, datatype, len);
+#endif
+
+    *error_code = MPI_SUCCESS;
+}
+
+void ADIOI_DAOS_ReadContig(ADIO_File fd, void *buf, int count,
+                           MPI_Datatype datatype, int file_ptr_type,
+                           ADIO_Offset offset, ADIO_Status * status, int *error_code)
+{
+    DAOS_IOContig(fd, buf, count, datatype, file_ptr_type,
+                  offset, status, NULL, DAOS_READ, error_code);
+}
+
+void ADIOI_DAOS_WriteContig(ADIO_File fd, const void *buf, int count,
+                            MPI_Datatype datatype, int file_ptr_type,
+                            ADIO_Offset offset, ADIO_Status * status, int *error_code)
+{
+    DAOS_IOContig(fd, (void *) buf, count, datatype, file_ptr_type,
+                  offset, status, NULL, DAOS_WRITE, error_code);
+}
+
+void ADIOI_DAOS_IReadContig(ADIO_File fd, void *buf, int count,
+                            MPI_Datatype datatype, int file_ptr_type,
+                            ADIO_Offset offset, MPI_Request * request, int *error_code)
+{
+    DAOS_IOContig(fd, buf, count, datatype, file_ptr_type,
+                  offset, NULL, request, DAOS_READ, error_code);
+}
+
+void ADIOI_DAOS_IWriteContig(ADIO_File fd, const void *buf, int count,
+                             MPI_Datatype datatype, int file_ptr_type,
+                             ADIO_Offset offset, MPI_Request * request, int *error_code)
+{
+    DAOS_IOContig(fd, (void *) buf, count, datatype, file_ptr_type,
+                  offset, NULL, request, DAOS_WRITE, error_code);
+}
+
```

Comment:
I like that you've combined the blocking and non-blocking code paths into a single common code block.   (No changes requested: just complimenting)